File: src/algorithms/infix-to-postfix-conversion-algorithm.py
def is_operator(ch):
    return ch in ['+', '-', '*', '/']

def precedence(ch):
    if ch in ['+', '-']:
        return 1
    if ch in ['*', '/']:
        return 2
    return 0

def infix_to_postfix(infix):
    postfix = []
    stack = []

    for ch in infix:
        if ch.isalnum():
            postfix.append(ch)
        elif ch == '(':
            stack.append(ch)
        elif ch == ')':
            while stack and stack[-1] != '(':
                postfix.append(stack.pop())
            if stack and stack[-1] == '(':
                stack.pop()
        elif is_operator(ch):
            while stack and precedence(stack[-1]) >= precedence(ch):
                postfix.append(stack.pop())
            stack.append(ch)

    while stack:
        postfix.append(stack.pop())

    return ''.join(postfix)

if __name__ == "__main__":
    infix_expression = "a+b*(c-d)/e"
    postfix_expression = infix_to_postfix(infix_expression)
    print(f"Infix expression: {infix_expression}")
    print(f"Postfix expression: {postfix_expression}")

File: src/algorithms/infix-to-postfix-conversion-algorithm.cpp
#include <iostream>
#include <stack>

using namespace std;

bool isOperator(char ch) {
    return (ch == '+' || ch == '-' || ch == '*' || ch == '/');
}

int precedence(char ch) {
    if (ch == '+' || ch == '-')
        return 1;
    if (ch == '*' || ch == '/')
        return 2;
    return 0;
}

void infixToPostfix(string infix, string &postfix) {
    stack<char> st;

    for (char &ch : infix) {
        if (isalnum(ch)) {
            postfix += ch;
        } else if (ch == '(') {
            st.push(ch);
        } else if (ch == ')') {
            while (!st.empty() && st.top() != '(') {
                postfix += st.top();
                st.pop();
            }
            if (!st.empty() && st.top() == '(') {
                st.pop();
            }
        } else if (isOperator(ch)) {
            while (!            st.empty() && precedence(st.top()) >= precedence(ch)) {
                postfix += st.top();
                st.pop();
            }
            st.push(ch);
        }
    }

    while (!st.empty()) {
        postfix += st.top();
        st.pop();
    }
}

int main() {
    string infix, postfix;

    cout << "Enter an infix expression: ";
    getline(cin, infix);

    infixToPostfix(infix, postfix);

    cout << "Postfix expression: " << postfix << endl;

    return 0;
}
